Add Google Cloud Workflows Operators

turbaszek opened the pull request:
Todo:
- [x] add sensor
- [x] add example dag
- [x] add system test
- [x] add docs

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

turbaszek committed:
Add Google Cloud Workflows Operators

turbaszek committed:
Add system test, example and sensor

fixup! Add system test, example and sensor

tolkjen commented on airflow/providers/google/cloud/operators/workflows.py:
```diff
@@ -0,0 +1,638 @@
+            metadata=self.metadata,
+        )
+        operation.result()
+        return Workflow.to_dict(operation)
```
Will this return a representation of the Workflow or the Long Running Operation?

turbaszek replied:
This will return a workflow

tolkjen replied:
Ok

mik-laj replied:
I am not sure if this is the correct code. It looks a little different in other places.
https://github.com/apache/airflow/blob/6570df871354f6bdefce88f21099fb09add096a4/airflow/providers/google/cloud/operators/dataproc.py#L546-L548

turbaszek replied:
Possibly, I will verify this once I'm able to run e2e tests - currently `operation.result()` does not work.

turbaszek replied:
Fixed

tolkjen commented on airflow/providers/google/cloud/operators/workflows.py:
```diff
@@ -0,0 +1,638 @@
+            timeout=self.timeout,
+            metadata=self.metadata,
+        )
+        return Execution.to_dict(execution)
```
The code should return only after the Execution isn't in "ACTIVE" state anymore:
https://cloud.google.com/workflows/docs/reference/executions/rest/v1beta/projects.locations.workflows.executions#State

turbaszek replied:
This operator only triggers an execution. Users should use sensor to wait for execution completion. Of course we can add synchronous waiting but this may influence idempotency

tolkjen replied:
Ok, non-blocking WorkflowExecutionsCreateExecutionOperator + sensor sound good.

tolkjen commented on airflow/providers/google/cloud/operators/workflows.py:
```diff
@@ -0,0 +1,638 @@
+            timeout=self.timeout,
+            metadata=self.metadata,
+        )
+        return [Execution.to_dict(e) for e in execution_iter]
```
The number of Executions returned can get pretty big (millions), so let's limit to executions started in the past 60 minutes. The LIST API doesn't support filtering today, so you'll need to do it yourself. You can use Execution.startTime for it:
https://cloud.google.com/workflows/docs/reference/executions/rest/v1beta/projects.locations.workflows.executions#State

turbaszek replied:
I will make it configurable then 👍

turbaszek replied:
Done

tolkjen reviewed (commented):
Thanks for implementing the operators so quickly!

turbaszek committed:
Rename operators and add docs

fixup! Rename operators and add docs

Apply suggestions from code review

Co-authored-by: Tobiasz Kędzierski <tobiasz.kedzierski@polidea.com>

Update airflow/providers/google/cloud/hooks/workflows.py

Co-authored-by: Tobiasz Kędzierski <tobiasz.kedzierski@polidea.com>

fixup! fixup! Rename operators and add docs

fixup! fixup! fixup! Rename operators and add docs

fixup! fixup! fixup! fixup! Rename operators and add docs

fixup! fixup! fixup! fixup! fixup! Rename operators and add docs

fixup! fixup! fixup! fixup! fixup! fixup! Rename operators and add docs

Update airflow/providers/google/cloud/example_dags/example_workflows.py

Co-authored-by: Tobiasz Kędzierski <tobiasz.kedzierski@polidea.com>

tolkjen commented on airflow/providers/google/cloud/sensors/workflows.py:
```diff
@@ -0,0 +1,123 @@
+        if state in self.failure_states:
+            raise AirflowException(
+                f"Execution {self.execution_id} for workflow {self.execution_id} "
+                f"failed and is in `{state}` state",
```
This exception will be raised also when the state is CANCELLED. For that reason I'd change the message to:
```suggestion
                f"finished in `{state}` state",
```

tolkjen commented on airflow/providers/google/cloud/sensors/workflows.py:
```diff
@@ -0,0 +1,123 @@
+            return True
+
+        self.log.info(
+            "Execution %s for workflow %s does not completed yet, current state: %s",
```
```suggestion
            "Execution %s for workflow %s hasn't completed yet, current state: %s",
```

TobKed commented on airflow/providers/google/cloud/example_dags/example_workflows.py:
```diff
@@ -0,0 +1,192 @@
+from airflow.providers.google.cloud.sensors.workflows import WorkflowExecutionSensor
+from airflow.utils.dates import days_ago
+
+LOCATION = os.environ.get("GCP_LOCATION", "us-central1")
```
```suggestion
LOCATION = os.environ.get("GCP_WORKFLOWS_LOCATION", "us-central1")
```
IMHO adding service name will make it more explicit and allow distinguishing env variables for other system tests

TobKed commented on airflow/providers/google/cloud/example_dags/example_workflows.py:
```diff
@@ -0,0 +1,192 @@
+LOCATION = os.environ.get("GCP_LOCATION", "us-central1")
+PROJECT_ID = os.environ.get("GCP_PROJECT_ID", "an-id")
+
+WORKFLOW_ID = "airflow-test-workflow"
```
```suggestion
WORKFLOW_ID = os.getenv("GCP_WORKFLOWS_WORKFLOW_ID", "airflow-test-workflow")
```
What do you think about using environmental variables for values which may collide if run within one project?
Environmental variables will allow to generate them in programmatic way which may allow for running multiple system tests within one project.

TobKed commented on airflow/providers/google/cloud/example_dags/example_workflows.py:
```diff
@@ -0,0 +1,192 @@
+    # [END how_to_create_workflow]
+
+    # [START how_to_update_workflows]
+    update_workflows = WorkflowsUpdateWorkflowOperator(
```
WDYT about adding value for `update_mask` parameter? It may be very helpful for new Workflows user to see how example value looks alike.

turbaszek replied:
Added 👌

TobKed commented on airflow/providers/google/cloud/operators/workflows.py:
```diff
@@ -0,0 +1,714 @@
+    """
+
+    template_fields = ("workflow_id",)
+    template_fields_renderers = {"workflow": "json"}
```
`workflow` is not listed in arguments. Shouldn't be supported alongside `update_mask`?

TobKed commented on airflow/providers/google/cloud/hooks/workflows.py:
```diff
@@ -0,0 +1,401 @@
+        Creates a new execution using the latest revision of
+        the given workflow.
+
+        :param execution: Required. Execution to be created.
```
WDYT about extending information about execution?
```suggestion
        :param execution: Required. Input parameters of the execution represented as a dictionary.
```

https://cloud.google.com/workflows/docs/reference/executions/rest/v1/projects.locations.workflows.executions#Execution

TobKed reviewed (changes requested):
Very nice! Good job!. I just left some minor comments

TobKed commented on airflow/providers/google/cloud/example_dags/example_workflows.py:
```diff
@@ -0,0 +1,197 @@
+
+EXECUTION = {"argument": ""}
+
+SLEEP_WORKFLOW_ID = "sleep_workflow"
```
```suggestion
SLEEP_WORKFLOW_ID = os.getenv("GCP_WORKFLOWS_SLEEP_WORKFLOW_ID", "sleep_workflow")
```
